pkg/kapp/diff/change_set_with_versioned_rs.go

```diff
@@ -254,44 +270,87 @@ func (d ChangeSetWithVersionedRs) newChange(existingRes, newRes ctlres.Resource)
 }
 
 type versionedResources struct {
-	Versioned    []ctlres.Resource
+	Versioned    []VersionedResource
 	NonVersioned []ctlres.Resource
 }
 
-func newVersionedResources(rs []ctlres.Resource) versionedResources {
-	var result versionedResources
-	for _, res := range rs {
-		_, hasVersionedAnn := res.Annotations()[versionedResAnnKey]
-		_, hasVersionedOrigAnn := res.Annotations()[versionedResOrigAnnKey]
+func (d versionedResources) VersionedRs() (rs []ctlres.Resource) {
+	for _, vres := range d.Versioned {
+		rs = append(rs, vres.Res())
+	}
+	return
+}
 
-		if hasVersionedAnn {
-			result.Versioned = append(result.Versioned, res)
-			if hasVersionedOrigAnn {
-				result.NonVersioned = append(result.NonVersioned, res.DeepCopy())
-			}
-		} else {
-			result.NonVersioned = append(result.NonVersioned, res)
-		}
+func (d versionedResources) NonVersionedRs() []ctlres.Resource {
+	return d.NonVersioned
+}
+
+func (d ChangeSetWithVersionedRs) newVersionedResources() versionedResources {
+	var result versionedResources
+	for _, res := range d.newRs {
+		result.appendRes(d.VersionedFromNewResource(res))
 	}
 	return result
 }
 
-func existingVersionedResources(rs []ctlres.Resource) versionedResources {
+func (d ChangeSetWithVersionedRs) existingVersionedResources() versionedResources {
 	var result versionedResources
-	for _, res := range rs {
-		// Expect that versioned resources should not be transient
-		// (Annotations may have been copied from versioned resources
-		// onto transient resources for non-versioning related purposes).
-		_, hasVersionedAnn := res.Annotations()[versionedResAnnKey]
+	for _, res := range d.existingRs {
+		result.appendRes(d.VersionedFromExistingResource(res))
```

Comment:
I think we can keep the logic here itself instead of a separate method. Also, using a public method inside a private method _generally_ is not a good idea. It usually leads to some issues (not in every case though).